katdal/dataset.py

```diff
@@ -190,6 +210,46 @@ def __hash__(self):
         """Base hash on description tuple, just like equality operator."""
         return hash(self._description)
 
+    def subset(self, first, last):
+        """Get a new :class:`SpectralWindow` representing a subset of the channels.
+
+        The returned :class:`SpectralWindow` covers the same frequencies as
+        channels [first, last) of the original.
+
+        Raises
+        ------
+        IndexError
+            If [first, last) is not a (non-empty) subinterval of the channels
+        """
+        if not (0 <= first < last <= self.num_chans):
+            raise IndexError('indices out of range')
+        channel_shift = (first + last) // 2 - self.num_chans // 2
```

Comment:
In the old `katsdpingest` code you mention:

```
The implementation is careful to avoid introducing rounding errors where the answer is an exactly representable value.
```

Is this still true? (It seems to be)

Reply:
Yes and no. The new functions avoid introducing any rounding errors, but when constructing a SpectralWindow you have to compute channel_width as `bandwidth / num_chans` which can introduce rounding errors.

One idea I've had is to give the SpectralWindow constructor an optional `bandwidth=` arg, which supersedes channel_width. That would allow any calculations regarding bandwidth to use the exact bandwidth, rather than reconstituting it from the (possibly inexact) channel width. It's a somewhat theoretical point, since CBF always have power-of-two channel counts and hence the division is always exact.

Reply:
I was going to make a similar suggestion... I'd be even more blunt and just change the API to expect a `bandwidth` parameter, since I don't *think* anyone else ever instantiated a `SpectralWindow` explicitly...

Reply:
I think @sjperkins would disagree: https://github.com/ska-sa/katsdppipelines/blob/0f594dd0a3962e880bc042e879b5c68692308b6c/katsdpcontim/katacomb/katacomb/mock_dataset.py#L369.

But since we both think the concept is good I'll work on my original suggestion.

Reply:
Ah, forgot to check that power user :-)

Reply:
I've implemented something and pushed it, but it's impossible to exactly guarantee all the desirable properties at once (in particular, subrange should ideally preserve channel_width and the frequencies of the selected channels, while rechannelise should preserve bandwidth), so I'm not sure if it's really a win that justifies the extra code complexity. Let me know what you think.